### DISCUSSION TITLE
Nested providers for subtree state

### DISCUSSION THREAD

**Discussion by timscullin** [pmndrs/jotai#682] [Q&A]
In the docs it says: "Multiple Providers can be used for multiple subtrees, even nested"
How do I handle nested providers? It doesn't seem like the subtrees can access the global/parent state.

In this example, how would the child counters display the parent count?

Thanks!

https://codesandbox.io/s/epic-pasteur-ky9ty?file=/src/App.js

```javascript
import "./styles.css";

import { atom, useAtom, Provider } from "jotai";

export default function App() {
  return (
    <>
      <Provider>
        <ParentCountItem />
        <Provider>
          <ChildCountItem />
        </Provider>
        <Provider>
          <ChildCountItem />
        </Provider>
      </Provider>
    </>
  );
}

const parentCountAtom = atom(0);
function useParentCount() {
  const [count, setCount] = useAtom(parentCountAtom);
  return { count, setCount };
}

function ParentCountItem() {
  const { count, setCount } = useParentCount();
  return (
    <div>
      <div>
        <strong>Parent</strong>
      </div>
      <div>Parent Count: {count}</div>
      <button
        onClick={() => {
          setCount(count + 1);
        }}
      >
        Increment
      </button>
    </div>
  );
}

const childCountAtom = atom(0);
function useChildCount() {
  const [count, setCount] = useAtom(childCountAtom);
  return { count, setCount };
}
function ChildCountItem() {
  const { count, setCount } = useChildCount();
  const { parentCount } = useParentCount();
  return (
    <div>
      <div>
        <strong>Child</strong>
      </div>
      <div>Parent Count: {parentCount || "Not Available"}</div>
      <div>Count: {count}</div>
      <button
        onClick={() => {
          setCount(count + 1);
        }}
      >
        Increment
      </button>
    </div>
  );
}

```

**Comment by dai-shi** [comment]
Docs may be misleading.
`useAtom` just looks for the closest provider, and it can't access to parent providers.

In this specific example, `scope` in Provider and useAtom solve the issue, but it might not be what you want. It's just like how the normal React Context works. You can just access to the closet provider in the tree.

**Reply by timscullin** [reply]
@dai-shi thanks! Makes total sense.
Nice to compose state with scope.

I implemented it below with child scopes and parent scopes.
https://codesandbox.io/s/child-scopes-omt09
https://codesandbox.io/s/parent-scope-0wvom

**Reply by Ekimoth** [reply]
@timscullin In your first example I noticed you're using two different scope strings (`childScope1` and `childScope2`) for your sibling `Provider`s - by doing that you're unnecessarily creating two contexts in runtime (or presumably way more if the number of sibling `Provider`s is dynamic in your actual project) while you only need one scope value (i.e. `childScope`), that is one context. Let me remind you that by rendering `n` sibling instances of a single `Context.Provider` component you've created that same number of mutually independent subtrees of the same shape - each child and every hook call down its subtree matching the scope value will only communicate with its own wrapping Provider with that same scope value.

**Comment by loganvolkers** [comment]
@timscullin we suffered from a similar problem in https://github.com/pmndrs/jotai/issues/783 and solved it with a separate package for creating atoms https://github.com/saasquatch/jotai-molecules

**Comment by bigdrum** [comment]
I find the provider can break the encapsulation more than the react context. With react context provider, it must specify which one to override. But Atom Provider overrides all values. This makes the Provider not quite composable.

Is it possible to provide a feature so that Provider can be specified a list of atoms that it tries to provide instead of all atoms? So that other atoms from the parent can still be propagated down.

(I know scope can handle that, but scope is very verbose and error prone, since people can forget. For many cases, we don't need that.)

**Reply by dai-shi** [reply]
What React Context can is `scope`. This is asking more than that. We know this is the limitation of how we use React Context, or what React Context alone provides. I agree that `scope` is not for apps.
We had a similar discussion before, and so far one of best approach is [jotai-molecules](https://github.com/saasquatch/jotai-molecules).

**Comment by methyl** [comment]
For what it's worth, I have another use case for nested providers. The idea is that I need an atom for which I want to get value from the closest provider, but still have normal access to global state. 

My current codebase solves it with React Context, but since I'm exploring Jotai, I wonder if there is any better answer than falling back to Context API.

[Here's what I figured out is possible as of today](https://codesandbox.io/s/elated-ives-d8mf2h?file=/src/App.js). Note that we can use the atom itself as the scope, as long as we keep provider to just single atom. The main downside is having to use special hook `useScopedAtom` and that it's not trivial to handle more than one atom provided.

```ts
import { atom, Provider, useAtom } from "jotai";

const scopedAtom = atom("");
const globalAtom = atom("test");

const useScopedAtom = (atom) => useAtom(atom, scopedAtom);

const ShowScopedValue = () => {
  const [global] = useAtom(globalAtom);
  const [scoped] = useScopedAtom(scopedAtom);
  return (
    <div>
      {scoped} {global}
    </div>
  );
};

const ScopedProvider = ({ atom, value, children }) => {
  return (
    <Provider initialValues={[[atom, value]]} scope={scopedAtom}>
      {children}
    </Provider>
  );
};

export default function App() {
  return (
    <>
      <Provider>
        <ScopedProvider atom={scopedAtom} value="outer">
          <ShowScopedValue />
          <ScopedProvider atom={scopedAtom} value="inner">
            <ShowScopedValue />
          </ScopedProvider>
        </ScopedProvider>
      </Provider>
    </>
  );
}


```

### Alternative API proposal

Not knowing a lot about Jotai internals yet, I want to humbly propose alternative API that simplifies the implementation. I hope it's not completely nuts.

```ts
import {atom, useAtom, Provider} from 'jotai'
const globalAtom = atom('')
const scopedAtom = atom('')

const ShowScopedValueMaybeSomeDay = () => {
  const [global] = useAtom(globalAtom);
  // First difference, I can just use plain useAtom.
  // The Provider decides whether it catches it or propagates up to the global provider
  const [scoped] = useAtom(scopedAtom);
  return (
    <div>
      {scoped} {global}
    </div>
  );
};

const ScopedProviderMaybeSomeDay = ({ value, children }) => {
  return (
    // Second difference, it's much easier to scope multiple atoms.
    // In currently working example it's not trivial, although I never attempted to figure it out.
    // Third difference, we could combine initialValues with providedAtoms definition to keep it short
    // as it's identical.
    <Provider providedAtoms={[[scopedAtom, value]]}>
      {children}
    </Provider>
  );
};
```

### Alternative API proposal, no changes to the core

I'm not sure if it's feasible to make `useAtom` transparently working with scopes, if not, I think something like this could work, possibly without touching the core.


```ts
import {atom, useAtom} from 'jotai'
import {atomScoped, useScopedAtom, ScopedProvider} from 'jotai/utils'

const globalAtom = atom('')
const scopedAtom = atomScoped('')


const ShowScopedValueMaybeSomeDay = () => {
  const [global] = useAtom(globalAtom);
  const [scoped] = useAtom(scopedAtom);

  return (
    <div>
      {scoped} {global}
    </div>
  );
};

const ScopedProviderMaybeSomeDay = ({ value, children }) => {
  return (
    <ScopedProvider providedAtoms={[[scopedAtom, value]]}>
      {children}
    </ScopedProvider>
  );
};
```

**Reply by dai-shi** [reply]
Have you looked jotai-molecules? Does it cover your use case?

**Reply by dai-shi** [reply]
I like the last example without touching the core.

**Reply by methyl** [reply]
Yes, jotai-molecules looks great but  for this simple usecase I think it's an overkill to pull in a dependency, if you can solve it with plain react contexts. 

Would you consider accepting PR with utilities adding similar API to the above?

**Reply by dai-shi** [reply]
Unfortunately, we would like keep the core as small as possible. It would be a field for third party libraries. 

There's a little possibility to add something in jotai/utils depending on its necessity _and_ simplicity, but can't tell for now.

Btw, I'm thinking about promoting some third party libraries somehow (semi?)officially. We need to build ecosystems.

**Reply by methyl** [reply]
Sure, I thought about extending `/utils`. I'll keep you posted.

**Reply by loganvolkers** [reply]
>Yes, jotai-molecules looks great but for this simple usecase I think it's an overkill to pull in a dependency, if you can solve it with plain react contexts.

@methyl `jotai-molecules` is dependency free, simple and based on react context. There is far more code in the test suite than the implementation (and a big part is based on `weakCache.ts` which is from `/utils`). I think it solves all your needs except:

1) It's not in `/utils`
2) You didn't write it

Maybe the best solution is for you to port it into `/utils`? I think that would solve 1 and 2.

**Reply by methyl** [reply]
Not sure, the API of `jotai-molecules` has bigger surface than what I proposed and introduces a lot of new concepts. It probably because it solves much more than what myself and OP need – that is, I want to control if a particular atom will hold its value in the closest scoped provider or in closest non-scoped provider. I wonder if such simpler API would solve at least some of the usecases for `jotai-molecules`.

**Reply by dai-shi** [reply]
I'm not yet convinced to add anything in jotai/utils, but if there were something, it would be nice if such a util can be a building block for jotai-molecules.
